boringtun/src/device/mod.rs

```diff
@@ -793,20 +998,61 @@ impl<T: Tun, S: Sock> Device<T, S> {
                         None => continue,
                     };
 
+                    if let Some(callback) = &d.config.firewall_process_outbound_callback {
+                        if !callback(&peer.tunnel.peer_static_public().to_bytes(), src) {
+                            continue;
+                        }
+                    }
+
+                    let mut public_key = String::with_capacity(32);
+                    for byte in peer.tunnel.peer_static_public().as_bytes() {
+                        let pub_symbol = format!("{:02X}", byte);
+                        public_key.push_str(&pub_symbol);
+                    }
+
                     match peer.tunnel.encapsulate(src, &mut t.dst_buf[..]) {
                         TunnResult::Done => {}
-                        TunnResult::Err(e) => error!(d.config.logger, "Encapsulate error {:?}", e),
+                        TunnResult::Err(e) => {
+                            tracing::error!(message = "Encapsulate error",
+                            error = ?e,
+                            public_key = public_key)
+                        }
                         TunnResult::WriteToNetwork(packet) => {
                             let endpoint = peer.endpoint();
                             if let Some(ref conn) = endpoint.conn {
+                                let addr = endpoint.addr;
                                 // Prefer to send using the connected socket
-                                conn.write(packet);
+                                if conn.write(packet) == 0 && packet.len() > 0 {
+                                    drop(endpoint);
+                                    peer.shutdown_endpoint();
+                                } else {
```

Comment:
This should also be trace, to avoid spam

Reply:
Done